File: tests/.hooks/csharp/custom-http/src/Speakeasy/CustomHttp/Hooks/CustomSecurityHook.cs
#nullable enable
namespace Speakeasy.CustomHttp.Hooks
{
    using System;
    using System.Net.Http;
    using System.Text.Json;
    using System.Threading.Tasks;
    using Speakeasy.CustomHttp.Models.Components;

    public class CustomSecurityHook : IBeforeRequestHook
    {
        public async Task<HttpRequestMessage> BeforeRequestAsync(BeforeRequestContext hookCtx, HttpRequestMessage request)
        {
            await Task.CompletedTask;

            switch (hookCtx.OperationID)
            {
                case "customHttpOnly":
                    if (hookCtx.SecuritySource == null)
                    {
                        throw new Exception("Security source is null");
                    }

                    Security? security = hookCtx.SecuritySource() as Security;
                    if (security == null)
                    {
                        throw new Exception("Security source is not of type Security");
                    }

                    if (security.CustomHttp == null)
                    {
                        throw new Exception("CustomHttp security is not defined");
                    }

                    var customHttp = security.CustomHttp;

                    request.Headers.Add("X-Security-UserID", customHttp.UserID.ToString());
                    request.Headers.Add("X-Security-Role", customHttp.Role.ToString().ToLower());
                    request.Headers.Add("X-Security-Passphrase", customHttp.Passphrase);
                    request.Headers.Add("X-Security-AccessCode", customHttp.AccessCode.ToString());

                    if (customHttp.Scopes != null && customHttp.Scopes.Count > 0)
                    {
                        var scopesJson = JsonSerializer.Serialize(customHttp.Scopes);
                        request.Headers.Add("X-Security-Scopes", scopesJson);
                    }

                    break;
            }

            return request;
        }
    }
}

File: tests/.hooks/php/custom-http/src/Hooks/CustomSecurityHook.php
<?php

declare(strict_types=1);

namespace OpenAPI\OpenAPI\Hooks;

use OpenAPI\OpenAPI\Models\Components\Security;
use Psr\Http\Message\RequestInterface;

class CustomSecurityHook implements BeforeRequestHook
{
    public function beforeRequest(BeforeRequestContext $context, RequestInterface $request): RequestInterface
    {
        switch ($context->operationID) {
            case 'customHttpOnly':
                if ($context->securitySource === null) {
                    throw new \InvalidArgumentException('Security source is null');
                }

                $security = $context->securitySource->call($context);

                if (! $security instanceof Security) {
                    throw new \InvalidArgumentException('Security source is not of type Security');
                }

                if ($security->customHttp === null) {
                    throw new \RuntimeException('CustomHttp security is not defined');
                }

                $customHttp = $security->customHttp;

                $request = $request->withAddedHeader('X-Security-UserID', (string) $customHttp->userID);
                $request = $request->withAddedHeader('X-Security-Role', $customHttp->role->value);
                $request = $request->withAddedHeader('X-Security-Passphrase', $customHttp->passphrase);

                if ($customHttp->accessCode !== null) {
                    $request = $request->withAddedHeader('X-Security-AccessCode', (string) $customHttp->accessCode);
                }

                if ($customHttp->scopes !== null && count($customHttp->scopes) > 0) {
                    $scopesJson = json_encode($customHttp->scopes);
                    if ($scopesJson === false) {
                        throw new \RuntimeException('Failed to encode scopes to JSON');
                    }
                    $request = $request->withAddedHeader('X-Security-Scopes', $scopesJson);
                }

                break;
        }

        return $request;
    }
}
